File: portfolio/cumulative_gains.go
package portfolio

import (
	"sort"

	decimal_opt "github.com/tsiemens/acb/decimal_value"
	"github.com/tsiemens/acb/util"
)

type CumulativeCapitalGains struct {
	CapitalGainsTotal      decimal_opt.DecimalOpt
	CapitalGainsYearTotals map[int]decimal_opt.DecimalOpt
}

func (g *CumulativeCapitalGains) CapitalGainsYearTotalsKeysSorted() []int {
	years := util.IntDecimalOptMapKeys(g.CapitalGainsYearTotals)
	sort.Ints(years)
	return years
}

func CalcSecurityCumulativeCapitalGains(deltas []*TxDelta) *CumulativeCapitalGains {
	var capGainsTotal decimal_opt.DecimalOpt
	capGainsYearTotals := util.NewDefaultMap(
		func(_ int) decimal_opt.DecimalOpt { return decimal_opt.Zero })

	for _, d := range deltas {
		if !d.CapitalGain.IsNull {
			capGainsTotal = capGainsTotal.Add(d.CapitalGain)
			yearTotalSoFar := capGainsYearTotals.Get(d.Tx.SettlementDate.Year())
			capGainsYearTotals.Set(d.Tx.SettlementDate.Year(), yearTotalSoFar.Add(d.CapitalGain))
		}
	}

	return &CumulativeCapitalGains{capGainsTotal, capGainsYearTotals.EjectMap()}
}

func CalcCumulativeCapitalGains(secGains map[string]*CumulativeCapitalGains) *CumulativeCapitalGains {
	var capGainsTotal decimal_opt.DecimalOpt
	capGainsYearTotals := util.NewDefaultMap(
		func(_ int) decimal_opt.DecimalOpt { return decimal_opt.Zero })

	for _, gains := range secGains {
		capGainsTotal = capGainsTotal.Add(gains.CapitalGainsTotal)
		for year, yearGains := range gains.CapitalGainsYearTotals {
			yearTotalSoFar := capGainsYearTotals.Get(year)
			capGainsYearTotals.Set(year, yearTotalSoFar.Add(yearGains))
		}
	}

	return &CumulativeCapitalGains{capGainsTotal, capGainsYearTotals.EjectMap()}
}

File: src/portfolio/cumulative_gains.rs
use std::collections::HashMap;

use rust_decimal::Decimal;

use super::{Security, TxDelta};

pub struct CumulativeCapitalGains {
    // None for registered affiliates (? maybe not).
    pub capital_gains_total: Decimal,
    pub capital_gains_years_totals: HashMap<i32, Decimal>,
}

impl CumulativeCapitalGains {
    pub fn capital_gains_year_totals_keys_sorted(&self) -> Vec<i32> {
        let mut years: Vec<i32> = self.capital_gains_years_totals.keys().copied().collect();
        years.sort();
        years
    }
}

pub fn calc_security_cumulative_capital_gains(deltas: &Vec<TxDelta>) -> CumulativeCapitalGains {
    let mut capital_gains_total = Decimal::ZERO;
    let mut cap_gains_year_totals = HashMap::<i32, Decimal>::new();

    for d in deltas {
        if let Some(cap_gain) = d.capital_gain {
            capital_gains_total += cap_gain;
            let year = d.tx.settlement_date.year();
            let year_total_so_far = cap_gains_year_totals.get(&year).unwrap_or(&Decimal::ZERO);
            cap_gains_year_totals.insert(year, year_total_so_far + cap_gain);
        }
    }

    CumulativeCapitalGains {
        capital_gains_total,
        capital_gains_years_totals: cap_gains_year_totals,
    }
}

pub fn calc_cumulative_capital_gains(sec_gains: &HashMap<Security, CumulativeCapitalGains>) -> CumulativeCapitalGains {
    let mut capital_gains_total = Decimal::ZERO;
    let mut cap_gains_year_totals = HashMap::<i32, Decimal>::new();

    for gains in sec_gains.values() {
        capital_gains_total += gains.capital_gains_total;
        for (year, year_gains) in &gains.capital_gains_years_totals {
            let year_total_so_far = cap_gains_year_totals.get(&year).unwrap_or(&Decimal::ZERO);
            cap_gains_year_totals.insert(*year, year_total_so_far + year_gains);
        }
    }

    CumulativeCapitalGains {
        capital_gains_total,
        capital_gains_years_totals: cap_gains_year_totals,
    }
}
